### DISCUSSION TITLE
Using MicroPython with 16MB Flash and 8MB PSRAM

### DISCUSSION THREAD

**Discussion by WolleKette** [micropython/micropython#10078] [Hardware & Peripherals] [answered]
I use the D32 Pro (V 2) Modul from WEMOS which has 16MB Flash and 8MB PSRAM. 
And I use a MicroPython Version with  support ffor SPIRAM from [here](https://micropython.org/download/esp32spiram/).

But I can’t see the Memory in MicroPython. Neither do I see the 16MB flash with `esp.flash_size()` nor the 8MB with `gc.mem_free()`. In both cases I only get 4MB displayed.

I think I have a fundamental problem understanding this. Is it at all possible to use more than 4MB heap RAM in MicroPython?

**Comment by davefes** [comment]
I copied these instructions off  a posting on the old forum.  Best to search the archives as I have not tried these myself.

*****
a) create a partition table for that 8 MB flash and place it in the esp32 directory. The name is arbitrary. I named it partitions_8mb.csv. My copy: 
Code: [Select all](https://forum.micropython.org/viewtopic.php?f=18&t=9768&start=10)
# Name,   Type, SubType, Offset,  Size, Flags
# Note: if you change the phy_init or app partition offset, make sure to change the offset in Kconfig.projbuild
nvs,      data, nvs,     0x9000,  0x6000,
phy_init, data, phy,     0xf000,  0x1000,
factory,  app,  factory, 0x10000, 0x1F0000,
vfs,      data, fat,     0x200000, 0x600000,
b) create a board definition subdirectory for that board. That could be a copy of the most similar board. I called it GENERIC_SPIRAM_8MB. But you can give it any suitable name. That name is used in the make command and in the changed mpconfigboard.cmake (see below).
c) create a file sdkconfig.board in the board directory which sets the flash size and partition table name. My version: 
Code: [Select all](https://forum.micropython.org/viewtopic.php?f=18&t=9768&start=10)
CONFIG_ESPTOOLPY_FLASHSIZE_8MB=y
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions_8mb.csv"
c) change the mpconfigboard.cmake file in adding the name of the sdkconfig.board file. Mine looks as follows after the change: 
Code: [Select all](https://forum.micropython.org/viewtopic.php?f=18&t=9768&start=10)
set(SDKCONFIG_DEFAULTS
    boards/sdkconfig.base
    boards/sdkconfig.ble
    boards/sdkconfig.spiram
    boards/GENERIC_SPIRAM_8MB/sdkconfig.board
)

set(MICROPY_FROZEN_MANIFEST ${MICROPY_BOARD_DIR}/manifest.py)
Then you can build & flash the image with 

make BOARD=<new_board_name> deploy

****

**Reply by WolleKette** [reply]
That sounds complicated. If I understand you correctly, I have to compile my own MicroPython image just to use more memory? Really?

**Comment by mattytrentini** [comment, marked as answer]
I'll address the RAM query. No, the ESP32 can only address 4MB of memory. Even with 8MB PSRAM fitted only 4MB of it will be accessible. It's why there's a number of 'red note alerts' on the [AliExpress Wemos page for the D32 Pro](https://www.aliexpress.com/item/32883116057.html):

![image](https://user-images.githubusercontent.com/194201/204004436-614bf12d-450d-48aa-90dd-ce6ee3e1c47d.png)

There actually _is_ a way to make some use of the additional memory but it's awkward, of limited use, and not supported in MicroPython. I mention it here only for completeness but I wouldn't suggest using it. Espressif provide the [himem allocation API](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/himem.html). It allows you to split your general memory, allocating a portion of it to be bank-switched at runtime. 

Himem is probably not worthwhile pursuing, especially with the newer ESP32-S2/S3 able to address more than 4MB RAM.

**Reply by WolleKette** [reply]
Thanks for the clarification!

**Comment by davefes** [comment]
Sorry I confused extending the RAM with extending the FLASH.
